core/src/main/java/org/apache/carbondata/core/metadata/SegmentFileStore.java

```diff
@@ -0,0 +1,690 @@
+        location = location.substring(tablePath.length(), location.length());
+        isRelative = true;
+      }
+      CarbonFile[] listFiles = carbonFile.listFiles(new CarbonFileFilter() {
+        @Override public boolean accept(CarbonFile file) {
+          return CarbonTablePath.isCarbonIndexFile(file.getAbsolutePath());
+        }
+      });
+      if (listFiles != null && listFiles.length > 0) {
+        SegmentFile localSegmentFile = new SegmentFile();
+        Map<String, FolderDetails> locationMap = new HashMap<>();
+        FolderDetails folderDetails = new FolderDetails();
+        folderDetails.setRelative(isRelative);
+        folderDetails.setPartitions(spec.getPartitions());
+        folderDetails.setStatus(SegmentStatus.SUCCESS.getMessage());
+        for (CarbonFile file : listFiles) {
+          if (file.getName().endsWith(CarbonTablePath.MERGE_INDEX_FILE_EXT)) {
+            folderDetails.setMergeFileName(file.getName());
+          } else {
+            folderDetails.getFiles().add(file.getName());
+          }
+        }
+        locationMap.put(location, folderDetails);
+        localSegmentFile.setLocationMap(locationMap);
+        if (segmentFile == null) {
+          segmentFile = localSegmentFile;
+        } else {
+          segmentFile = segmentFile.merge(localSegmentFile);
+        }
+      }
+    }
+    return segmentFile;
+  }
+
+  /**
+   * This method reads the segment file which is written in json format
+   *
+   * @param segmentFilePath
+   * @return
+   */
+  private SegmentFile readSegmentFile(String segmentFilePath) throws IOException {
+    Gson gsonObjectToRead = new Gson();
+    DataInputStream dataInputStream = null;
+    BufferedReader buffReader = null;
+    InputStreamReader inStream = null;
+    SegmentFile segmentFile;
+    AtomicFileOperations fileOperation =
+        new AtomicFileOperationsImpl(segmentFilePath, FileFactory.getFileType(segmentFilePath));
+
+    try {
+      if (!FileFactory.isFileExist(segmentFilePath, FileFactory.getFileType(segmentFilePath))) {
+        return null;
+      }
+      dataInputStream = fileOperation.openForRead();
+      inStream = new InputStreamReader(dataInputStream,
+          Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET));
+      buffReader = new BufferedReader(inStream);
+      segmentFile = gsonObjectToRead.fromJson(buffReader, SegmentFile.class);
+    } finally {
+      if (inStream != null) {
+        CarbonUtil.closeStreams(buffReader, inStream, dataInputStream);
+      }
+    }
+
+    return segmentFile;
+  }
+
+  /**
+   * Reads segment file.
+   */
+  public void readSegment(String tablePath, String segmentFileName) throws IOException {
+    String segmentFilePath =
+        CarbonTablePath.getSegmentFilesLocation(tablePath) + CarbonCommonConstants.FILE_SEPARATOR
+            + segmentFileName;
+    SegmentFile segmentFile = readSegmentFile(segmentFilePath);
+    this.tablePath = tablePath;
+    this.segmentFile = segmentFile;
```

Comment:
Here we are directly updating the member variable. If it is called from multiple places then in concurrent operations it can cause problem. Kindly re-verify and if possible avoid updating the member variables if getting called from multiple places

Reply:
Ok, moved to constructor